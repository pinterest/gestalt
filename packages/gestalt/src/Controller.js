// @flow
import * as React from 'react';
import PropTypes from 'prop-types';
import Contents from './Contents.js';
import OutsideEventBehavior from './behaviors/OutsideEventBehavior.js';

type Props = {|
  anchor: ?HTMLElement,
  bgColor: 'blue' | 'darkGray' | 'orange' | 'white',
  children?: React.Node,
  idealDirection?: 'up' | 'right' | 'down' | 'left',
  onDismiss: () => void,
  positionRelativeToAnchor: boolean,
  shouldFocus?: boolean,
  size?: 'xs' | 'sm' | 'md' | 'lg' | 'xl' | number,
|};

const SIZE_WIDTH_MAP = {
  xs: 185,
  sm: 230,
  md: 284,
  lg: 320,
  xl: 375,
};

const ESCAPE_KEY_CODE = 27;

type ClientRect = {
  bottom: number,
  height: number,
  left: number,
  right: number,
  top: number,
  width: number,
};

type State = {
  relativeOffset: {
    x: number,
    y: number,
  },
  triggerBoundingRect: ClientRect,
};

export default class Controller extends React.Component<Props, State> {
  state: State = {
    relativeOffset: {
      x: 0,
      y: 0,
    },
    triggerBoundingRect: {
      bottom: 0,
      height: 0,
      left: 0,
      right: 0,
      top: 0,
      width: 0,
    },
  };

  componentDidMount() {
    this.updateTriggerRect(this.props);
  }

  componentWillReceiveProps(nextProps: Props) {
    this.updateTriggerRect(nextProps);
  }

  handleKeyDown = (event: { keyCode: number }) => {
    if (event.keyCode === ESCAPE_KEY_CODE) {
      this.props.onDismiss();
    }
  };

  handlePageClick = (event: Event) => {
    if (
      event.target instanceof Node &&
      this.props.anchor &&
      !this.props.anchor.contains(event.target)
    ) {
      this.props.onDismiss();
    }
  };

  handleResize = () => {
    this.updateTriggerRect(this.props);
  };

  updateTriggerRect = (props: Props) => {
    const { anchor, positionRelativeToAnchor } = props;
    let triggerBoundingRect;
    let relativeOffset;
    if (anchor) {
      triggerBoundingRect = anchor.getBoundingClientRect();

      // Needed for correct positioning within Contents.js
      relativeOffset = {
        x: positionRelativeToAnchor
          ? triggerBoundingRect.left - anchor.offsetLeft
          : 0,
        y: positionRelativeToAnchor
          ? triggerBoundingRect.top - anchor.offsetTop
          : 0,
      };
    }

    this.setState({ relativeOffset, triggerBoundingRect });
  };

  render() {
    const {
      anchor,
      bgColor,
      children,
      idealDirection,
      positionRelativeToAnchor,
      shouldFocus,
    } = this.props;
    if (!anchor) {
      return null;
    }
    const size = this.props.size ? this.props.size : 'sm';
    const width = typeof size === 'string' ? SIZE_WIDTH_MAP[size] : size;
    return (
      <OutsideEventBehavior onClick={this.handlePageClick}>
        <Contents
          bgColor={bgColor}
          idealDirection={idealDirection}
          onKeyDown={this.handleKeyDown}
          onResize={this.handleResize}
          positionRelativeToAnchor={positionRelativeToAnchor}
          relativeOffset={this.state.relativeOffset}
          shouldFocus={shouldFocus}
          triggerRect={this.state.triggerBoundingRect}
          width={width}
        >
          {children}
        </Contents>
      </OutsideEventBehavior>
    );
  }
}

Controller.propTypes = {
  anchor: PropTypes.shape({
    contains: PropTypes.func,
    getBoundingClientRect: PropTypes.func,
  }),
  bgColor: PropTypes.oneOf(['blue', 'darkGray', 'orange', 'white']),
  children: PropTypes.node,
  idealDirection: PropTypes.oneOf(['up', 'right', 'down', 'left']),
  onDismiss: PropTypes.func.isRequired,
  positionRelativeToAnchor: PropTypes.bool,
  shouldFocus: PropTypes.bool,
  size: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl']), // default: sm
  ]),
};
