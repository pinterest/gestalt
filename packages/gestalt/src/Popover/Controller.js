// @flow strict
import { type Node as ReactNode } from 'react';
import { FloatingPortal } from '@floating-ui/react';
import Contents, { type Role } from './Contents';
import OutsideEventBehavior from '../behaviors/OutsideEventBehavior';
import { ESCAPE } from '../keyCodes';

const SIZE_WIDTH_MAP = {
  xs: 180,
  sm: 230,
  md: 284,
  lg: 320,
  xl: 360,
};

type Props = {
  accessibilityLabel?: string,
  anchor: HTMLElement,
  bgColor: 'blue' | 'darkGray' | 'white',
  border?: boolean,
  caret?: boolean,
  children?: ReactNode,
  onKeyDown?: ({| event: SyntheticKeyboardEvent<HTMLElement> |}) => void,
  id?: ?string,
  idealDirection?: 'up' | 'right' | 'down' | 'left' | 'forceDown',
  onDismiss: () => void,
  role?: ?Role,
  rounding?: 2 | 4,
  shouldFocus?: boolean,
  size?: 'xs' | 'sm' | 'md' | 'lg' | 'xl' | number | null,
  disablePortal?: boolean,
  scrollBoundary?: HTMLElement,
  hideWhenReferenceHidden?: boolean,
  onPositioned?: () => void,
  shouldTrapFocus?: boolean,
};

export default function Controller({
  accessibilityLabel,
  anchor,
  bgColor,
  border,
  caret,
  children,
  id,
  idealDirection,
  role,
  rounding,
  shouldFocus,
  size = 'sm',
  onDismiss,
  onKeyDown,
  disablePortal,
  scrollBoundary,
  hideWhenReferenceHidden,
  onPositioned,
  shouldTrapFocus,
}: Props): ReactNode {
  const width = typeof size === 'string' ? SIZE_WIDTH_MAP[size] : size;

  const handleKeyDown = (event: SyntheticKeyboardEvent<HTMLElement>) => {
    if (event.keyCode === ESCAPE) {
      onDismiss();
    }
    onKeyDown?.({ event });
  };

  const handlePageClick = (event: Event) => {
    if (event.target instanceof Node && !anchor.contains(event.target)) {
      onDismiss();
    }
  };

  const contents = (
    <OutsideEventBehavior onClick={handlePageClick}>
      <Contents
        accessibilityLabel={accessibilityLabel}
        anchor={anchor}
        bgColor={bgColor}
        border={border}
        caret={caret}
        hideWhenReferenceHidden={hideWhenReferenceHidden}
        id={id}
        idealDirection={idealDirection}
        onKeyDown={handleKeyDown}
        onPositioned={onPositioned}
        role={role}
        rounding={rounding}
        scrollBoundary={scrollBoundary}
        shouldFocus={shouldFocus}
        shouldTrapFocus={shouldTrapFocus}
        width={width}
      >
        {children}
      </Contents>
    </OutsideEventBehavior>
  );

  return disablePortal ? contents : <FloatingPortal>{contents}</FloatingPortal>;
}
