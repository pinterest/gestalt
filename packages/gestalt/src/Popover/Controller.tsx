import { ReactNode } from 'react';
import { FloatingPortal } from '@floating-ui/react';
import Contents, { Role } from './Contents';
import OutsideEventBehavior from '../behaviors/OutsideEventBehavior';
import { Overflow } from '../boxTypes';
import { ESCAPE } from '../keyCodes';

const SIZE_WIDTH_MAP = {
  xs: 180,
  sm: 230,
  md: 284,
  lg: 320,
  xl: 360,
} as const;

type Props = {
  accessibilityLabel?: string;
  anchor: HTMLElement;
  bgColor?: 'blue' | 'darkGray' | 'white';
  border?: boolean;
  caret?: boolean;
  children?: ReactNode;
  onKeyDown?: (arg1: { event: React.KeyboardEvent<HTMLElement> }) => void;
  id?: string;
  idealDirection?: 'up' | 'right' | 'down' | 'left' | 'forceDown';
  onDismiss: () => void;
  role?: Role;
  rounding?: 2 | 4;
  shouldFocus?: boolean;
  size?: 'xs' | 'sm' | 'md' | 'lg' | 'xl' | number;
  disablePortal?: boolean;
  scrollBoundary?: HTMLElement;
  hideWhenReferenceHidden?: boolean;
  onPositioned?: () => void;
  shouldTrapFocus?: boolean;
  overflow?: Extract<Overflow, 'auto' | 'hidden' | 'visible'>;
};

export default function Controller({
  accessibilityLabel,
  anchor,
  bgColor,
  border,
  caret,
  children,
  id,
  idealDirection,
  role,
  rounding,
  shouldFocus,
  size,
  onDismiss,
  onKeyDown,
  disablePortal,
  scrollBoundary,
  hideWhenReferenceHidden,
  onPositioned,
  shouldTrapFocus,
  overflow,
}: Props) {
  const width = typeof size === 'string' ? SIZE_WIDTH_MAP[size] : size;

  const handleKeyDown = (event: React.KeyboardEvent<HTMLElement>) => {
    if (event.keyCode === ESCAPE) {
      onDismiss();
    }
    onKeyDown?.({ event });
  };

  const handlePageClick = (event: Event) => {
    if (event.target instanceof Node && !anchor.contains(event.target)) {
      onDismiss();
    }
  };

  const contents = (
    <OutsideEventBehavior onClick={handlePageClick}>
      <Contents
        accessibilityLabel={accessibilityLabel}
        anchor={anchor}
        bgColor={bgColor}
        border={border}
        caret={caret}
        hideWhenReferenceHidden={hideWhenReferenceHidden}
        id={id}
        idealDirection={idealDirection}
        onKeyDown={handleKeyDown}
        onPositioned={onPositioned}
        overflow={overflow}
        role={role}
        rounding={rounding}
        scrollBoundary={scrollBoundary}
        shouldFocus={shouldFocus}
        shouldTrapFocus={shouldTrapFocus}
        width={width}
      >
        {children}
      </Contents>
    </OutsideEventBehavior>
  );

  return disablePortal ? contents : <FloatingPortal>{contents}</FloatingPortal>;
}
