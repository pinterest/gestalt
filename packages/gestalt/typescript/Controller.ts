import type { Node as ReactNode } from "react";
import { Component } from "react";
import { ESCAPE } from "./keyCodes";
import type { Role } from "./Contents";
import Contents from "./Contents";
import OutsideEventBehavior from "./behaviors/OutsideEventBehavior";
import { useScrollBoundaryContainer } from "./contexts/ScrollBoundaryContainer";
import type { ClientRect, Coordinates } from "./utils/positioningTypes";
import { getTriggerRect } from "./utils/positioningUtils";
const SIZE_WIDTH_MAP = {
  xs: 180,
  sm: 230,
  md: 284,
  lg: 320,
  xl: 360,
};
type OwnProps = {
  anchor: HTMLElement;
  bgColor: "blue" | "darkGray" | "orange" | "red" | "white";
  border?: boolean;
  caret?: boolean;
  children?: ReactNode;
  handleKeyDown?: (event: React.KeyboardEvent<HTMLElement>) => void;
  id?: string | null | undefined;
  idealDirection?: "up" | "right" | "down" | "left";
  onDismiss: () => void;
  positionRelativeToAnchor: boolean;
  role?: Role | null | undefined;
  rounding?: 2 | 4;
  shouldFocus?: boolean;
  size?: "xs" | "sm" | "md" | "lg" | "xl" | number | null;
};
type HookProps = {
  scrollBoundaryContainerRef: HTMLElement | null | undefined;
};
type Props = OwnProps & HookProps;
type State = {
  relativeOffset: Coordinates;
  triggerBoundingRect: ClientRect;
};

class Controller extends Component<Props, State> {
  static defaultProps: {
    size: "xs" | "sm" | "md" | "lg" | "xl" | number | null;
  } = {
    // Default size only applies when size is omitted,
    // if passed as null it will remain null
    size: "sm",
  };
  state = {
    relativeOffset: {
      x: 0,
      y: 0,
    },
    triggerBoundingRect: {
      bottom: 0,
      height: 0,
      left: 0,
      right: 0,
      top: 0,
      width: 0,
    },
  };

  static getDerivedStateFromProps({
    anchor,
    positionRelativeToAnchor,
    scrollBoundaryContainerRef,
  }: Props) {
    return getTriggerRect({
      anchor,
      positionRelativeToAnchor,
      scrollBoundaryContainerRef,
    });
  }

  componentDidMount() {
    this.updateTriggerRect(this.props);
  }

  handleKeyDown: (event: React.KeyboardEvent<HTMLElement>) => void = (
    event
  ) => {
    const { handleKeyDown, onDismiss } = this.props;

    if (event.keyCode === ESCAPE) {
      onDismiss();
    }

    if (handleKeyDown) handleKeyDown(event);
  };
  handlePageClick: (event: Event) => void = (event) => {
    const { anchor, onDismiss } = this.props;

    if (event.target instanceof Node && !anchor.contains(event.target)) {
      onDismiss();
    }
  };
  handleResize: () => void = () => {
    this.updateTriggerRect(this.props);
  };
  updateTriggerRect: (arg0: Props) => void = ({
    anchor,
    positionRelativeToAnchor,
    scrollBoundaryContainerRef,
  }: Props) => {
    const { relativeOffset, triggerBoundingRect } = getTriggerRect({
      anchor,
      positionRelativeToAnchor,
      scrollBoundaryContainerRef,
    });
    this.setState({
      relativeOffset,
      triggerBoundingRect,
    });
  };

  render(): ReactNode {
    const {
      anchor,
      bgColor,
      border,
      caret,
      children,
      id,
      idealDirection,
      positionRelativeToAnchor,
      role,
      rounding,
      shouldFocus,
      size,
    } = this.props;
    const { relativeOffset, triggerBoundingRect } = this.state;
    const width = typeof size === "string" ? SIZE_WIDTH_MAP[size] : size;
    return (
      <OutsideEventBehavior onClick={this.handlePageClick}>
        <Contents
          anchor={anchor}
          bgColor={bgColor}
          border={border}
          caret={caret}
          id={id}
          idealDirection={idealDirection}
          onKeyDown={this.handleKeyDown}
          onResize={this.handleResize}
          positionRelativeToAnchor={positionRelativeToAnchor}
          relativeOffset={relativeOffset}
          role={role}
          rounding={rounding}
          shouldFocus={shouldFocus}
          triggerRect={triggerBoundingRect}
          width={width}
        >
          {children}
        </Contents>
      </OutsideEventBehavior>
    );
  }
}

const WrappedController = (props: OwnProps): ReactNode => {
  const { scrollBoundaryContainerRef = null } = useScrollBoundaryContainer();
  return (
    <Controller
      {...props}
      scrollBoundaryContainerRef={scrollBoundaryContainerRef}
    />
  );
};

export default WrappedController;