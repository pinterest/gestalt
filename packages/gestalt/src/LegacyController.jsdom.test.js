// @flow strict
import { create } from 'react-test-renderer';
import Controller from './LegacyController';

test('Controller renders', () => {
  const element = document.createElement('div');
  const component = create(
    <Controller
      anchor={element}
      bgColor="darkGray"
      id=""
      onDismiss={() => {}}
      positionRelativeToAnchor
      role={undefined}
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
