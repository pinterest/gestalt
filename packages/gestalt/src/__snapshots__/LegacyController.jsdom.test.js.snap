// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controller renders 1`] = `
<div>
  <div
    className="container contents maxDimensions minDimensions"
    style={
      Object {
        "left": -10,
        "top": 4,
        "visibility": "visible",
        "zIndex": undefined,
      }
    }
    tabIndex={-1}
  >
    <div
      className="caretSecondary caret"
      style={
        Object {
          "bottom": null,
          "left": 10,
          "right": null,
          "top": -4,
        }
      }
    >
      <svg
        height={4}
        width={12}
      >
        <path
          d="M0 4c.694 0 1.36-.278 1.846-.773L4.376.66c.806-.819 2.187-.885 3.083-.148.057.047.111.096.163.148l2.526 2.565A2.6 2.6 0 0012 4"
          stroke="rgb(0 0 0 / 0.02)"
        />
      </svg>
    </div>
    <div
      className="border innerContents maxDimensions minDimensions secondary"
      id=""
      style={
        Object {
          "maxHeight": 691.1999999999999,
          "maxWidth": 230,
        }
      }
    />
  </div>
</div>
`;
